Maze.cpp

```diff
@@ -34,11 +38,194 @@ void Maze::setWall(CellCoordinate pos, Direction dir, bool wall) {
 }
 
 
-Path Maze::findPath(CellCoordinate start, CellCoordinate end)
+Path Maze::findPath(CellCoordinate start, CellCoordinate end, Direction facing )
+{
+	Node * startNode = getNode(start);
+
+	// The nodes that still need to be evaluated.
+	// Initially insert the start node.
+	set<Node *> openNodes;
+	openNodes.insert(getNode(start));
+	
+	// Reset any metadata from previous pathfinding.
+	resetNodePathData();
+	
+	// The start node is 0 distance away.
+	startNode->setScores(0, heuristic(start, end));
+	startNode->direction = facing;
+
+	// While there are node remaining to be evaluated.
+	while (!openNodes.empty())
+	{
+		// Get the open node with lowest score.
+		auto currentNodeItr = openNodes.begin();
+
+		// If the current node is the finish node.
+		if ( *currentNodeItr == getNode( end ) )
+		{
+			// We are done. Construct the path.
+			return constructPath(*currentNodeItr);
+		}
+		
+		// Set the current node as evaluated.
+		openNodes.erase(currentNodeItr);
+		(*currentNodeItr)->evaluated = true;
+
+		// For each node adjacent to the current node.
+		for (
+			Direction direction = Direction::N;
+			direction != Direction::NONE;
+#ifdef MAZE_DIAGONALS
+			direction = static_cast<Direction>(direction + 1)
+#else
+			direction = static_cast<Direction>(direction + 2)
+#endif // MAZE_DIAGONALS
+			)
+		{
+			Node * adjacentNode = getAdjacentNode(*currentNodeItr, direction);
+
+			// If the adjacent node has already been evaluated or does not exist.
+			if (adjacentNode->evaluated || !adjacentNode->exists)
+			{
+				// Ignore the adjacent node.
+				continue;
+			}
+
+			auto tentativeScore = (*currentNodeItr)->gScore + calculateMovementCost( (*currentNodeItr)->direction, direction );
+
+			// If adjacent node not in openNodes.
+			if (openNodes.find(adjacentNode) == openNodes.end() )
+			{
+				// Discover a new node.
+				openNodes.insert(adjacentNode);
+			}
+			// If the previous score given to the adjacent node is better.
+			else if (tentativeScore >= adjacentNode->gScore)
+			{
+				// This is not a better path.
+				continue;
+			}
+
+			// This path is the best so far.
+			adjacentNode->previous = *currentNodeItr;
+			auto heuristicScore = heuristic((*currentNodeItr)->pos, end);
+			(*currentNodeItr)->setScores(tentativeScore, adjacentNode->gScore + heuristicScore);
+		}
+	}
+
+	throw runtime_error("No path found");
+}
+
+Node * Maze::getNode(NodeCoordinate pos)
 {
+	return &maze[pos.y][pos.x];
+}
 
+Node * Maze::getAdjacentNode(Node * node, Direction direction)
```

Comment:
Yeah, it might be cryptic using `operator+` for this...

Anyway, wouldn't it be better if this was a method of `Node`? The node would need a pointer/reference to the `Maze`.